### ISSUE TITLE
Building 0.16 on i686 fails

### ISSUE THREAD

**Issue by magthe** [haskell-crypto/cryptonite#88]
The failed build ends with this bag of errors:

```
cbits/cryptonite_rdrand.c: In function ‘cryptonite_get_rand_bytes’:

cbits/cryptonite_rdrand.c:63:2: error:
     error: inconsistent operand constraints in an ‘asm’
      asm(".byte 0x48,0x0f,0xc7,0xf0; setc %1" \
      ^

cbits/cryptonite_rdrand.c:78:3: error:
     note: in expansion of macro ‘inline_rdrand_rax’
       inline_rdrand_rax(tmp, ok);
       ^~~~~~~~~~~~~~~~~

cbits/cryptonite_rdrand.c:63:2: error:
     error: inconsistent operand constraints in an ‘asm’
      asm(".byte 0x48,0x0f,0xc7,0xf0; setc %1" \
      ^

cbits/cryptonite_rdrand.c:87:3: error:
     note: in expansion of macro ‘inline_rdrand_rax’
       inline_rdrand_rax(tmp, ok);
       ^~~~~~~~~~~~~~~~~

cbits/cryptonite_rdrand.c:63:2: error:
     error: inconsistent operand constraints in an ‘asm’
      asm(".byte 0x48,0x0f,0xc7,0xf0; setc %1" \
      ^

cbits/cryptonite_rdrand.c:87:3: error:
     note: in expansion of macro ‘inline_rdrand_rax’
       inline_rdrand_rax(tmp, ok);
       ^~~~~~~~~~~~~~~~~

cbits/cryptonite_rdrand.c:63:2: error:
     error: inconsistent operand constraints in an ‘asm’
      asm(".byte 0x48,0x0f,0xc7,0xf0; setc %1" \
      ^

cbits/cryptonite_rdrand.c:94:3: error:
     note: in expansion of macro ‘inline_rdrand_rax’
       inline_rdrand_rax(tmp, ok);
       ^~~~~~~~~~~~~~~~~
`gcc' failed in phase `C Compiler'. (Exit code: 1)
```

I'm building on Arch Linux, i686 with

```
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.0.1
$ gcc --version
gcc (GCC) 6.1.1 20160501
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by domenkozar** [comment]
Same on NixOS: http://hydra.nixos.org/build/36573723/nixlog/1/raw

**Comment by centromere** [comment]
@magthe Please try building the [rdrand-fix](https://github.com/haskell-crypto/cryptonite/tree/rdrand-fix) branch.

**Comment by vincenthz** [comment]
I've added a commit on master f5a811e755eee253bc34897f327e096e04ad1286 that might fix the building on i686 without disabling rdrand altogether. if this fail then we should just disable it just like @centromere rdrand-fix branch does.

@magthe @domenkozar please test master

**Comment by huggablemonad** [comment]
I seem to be getting a related error that fails earlier:

```
cbits/cryptonite_rdrand.c: In function 'cryptonite_cpu_has_rdrand':

cbits/cryptonite_rdrand.c:40:2: error:
     error: inconsistent operand constraints in an 'asm'
      __asm__ volatile ("cpuid": "=a" (ax), "=b" (bx), "=c" (cx), "=d" (dx) : "a" (func));
      ^
```

I think it might be caused by the `-fPIC` flag.

```
test_rdrand.c:
#include <stdint.h>

int cryptonite_cpu_has_rdrand()
{
        uint32_t ax,bx,cx,dx,func=1;
        __asm__ volatile ("cpuid": "=a" (ax), "=b" (bx), "=c" (cx), "=d" (dx) : "a" (func));
        return (cx & 0x40000000);
}

$ gcc -c test_rdrand.c
$ # success
$ gcc -c -fPIC test_rdrand.c
rdrand.c: In function 'cryptonite_cpu_has_rdrand':
rdrand.c:6:2: error: inconsistent operand constraints in an 'asm'
  __asm__ volatile ("cpuid": "=a" (ax), "=b" (bx), "=c" (cx), "=d" (dx) : "a" (func));
  ^
```

Disabling rdrand allows the build to succeed.

I'm using GHC 8.0.1 and GCC 4.9.3 on Slackware Linux i686.

**Comment by magthe** [comment]
@vincenthz Your fix, https://github.com/haskell-crypto/cryptonite/commit/f5a811e755eee253bc34897f327e096e04ad1286, allows the build to succeed.

I ran the tests too and they all passed.

**Comment by andreyk0** [comment]
Tried compiling latest master (with the fix) for armhf platform, GHC 8.0.1, stack nightly-2016-06-12, still getting an asm error (below). 

```
$ stack test
cryptonite-0.16: build (lib + test)
Preprocessing library cryptonite-0.16...
cbits/cryptonite_rdrand.c: In function 'cryptonite_cpu_has_rdrand':

/home/andrey/work/cryptonite/cbits/cryptonite_rdrand.c:40:2: error:
     error: impossible constraint in 'asm'
      __asm__ volatile ("cpuid": "=a" (ax), "=b" (bx), "=c" (cx), "=d" (dx) : "a" (func));
      ^      
`gcc' failed in phase `C Compiler'. (Exit code: 1)
Progress: 1/2
--  While building package cryptonite-0.16 using:
      /home/andrey/.stack/setup-exe-cache/arm-linux/setup-Simple-Cabal-1.24.0.0-ghc-8.0.1 --builddir=.stack-work/dist/arm-linux/Cabal-1.24.0.0 build lib:cryptonite test:test-cryptonite --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1

```

**Comment by centromere** [comment]
Just a stab in the dark, does setting (or not setting) `-fPIC` change anything?

**Comment by andreyk0** [comment]
@centromere I can try (what's the right setting to temporarily alter in the cabal file?) but I don't think it'll make any difference for ARM platform.
Looking at the patch, it assumes that a platform that is not x86_64 is i686.
That's not going to work for ARM because rdrand is an intel-specific hardware feature.
Ideally this feature would only be enabled on the platforms that support it.

**Comment by centromere** [comment]
The setting would be `ghc-options`. You can try building with `-f-support_rdrand` too.

**Comment by andreyk0** [comment]
@centromere Thanks! Sorry, that didn't help, same error with fPIC option.
Disabling rdrand does help, all tests pass.
I think that's the right thing to do and a trivial change if others agree.

Here's how I ran it and all compilation/tests passed on x86_64 and on arm:

```
diff --git a/cryptonite.cabal b/cryptonite.cabal
index dcdfdf8..84547ee 100644
--- a/cryptonite.cabal
+++ b/cryptonite.cabal
@@ -241,7 +241,7 @@ Library
   if arch(x86_64)
     CPP-options: -DARCH_X86_64

-  if flag(support_rdrand)
+  if flag(support_rdrand) && (arch(x86_64) || arch(i686))
     CPP-options:    -DSUPPORT_RDRAND
     Other-modules:  Crypto.Random.Entropy.RDRand
     c-sources:      cbits/cryptonite_rdrand.c
~/work/cryptonite# 

```

**Comment by mitchty** [comment]
I can agree with @andreyk0 I could build cryptonite on arm prior to 0.16 and the rdrand stuff doesn't make much sense to try on arm so it should be gated to i386.

**Comment by centromere** [comment]
@andreyk0 @mitchty Check out 32aec08dbfe

**Comment by mitchty** [comment]
@centromere yep works fine then, ship it 👍

**Comment by andreyk0** [comment]
Thanks, guys! master looks good on arm, all tests pass!

**Comment by centromere** [comment]
@mitchty I do not have permission to do that. @vincenthz

**Comment by huggablemonad** [comment]
@vincenthz 8c19352e486ace8b73c68b2de745326a892dc680 fixes the `fPIC`-related error I was seeing. Thanks!

**Comment by vincenthz** [comment]
great @huggablemonad . complete stab in the dark, as I couldn't test it, so I'm really glad that you following up with success !

**Comment by vincenthz** [comment]
I think it's all fixed in 0.17 released yesterday. as usual everyone don't hesitate to give us a howl or/and open issue if needed
